src/test/epoll/test_epoll_edge.rs

```diff
@@ -0,0 +1,414 @@
+    socket_init(SocketInitMethod::UnixSocketpair, libc::SOCK_STREAM)
+}
+fn unix_pair_dgram_fds_init_helper() -> (libc::c_int, libc::c_int) {
+    socket_init(SocketInitMethod::UnixSocketpair, libc::SOCK_DGRAM)
+}
+fn unix_pair_seqpacket_fds_init_helper() -> (libc::c_int, libc::c_int) {
+    socket_init(SocketInitMethod::UnixSocketpair, libc::SOCK_SEQPACKET)
+}
+
+fn pipe_fds_init_helper() -> (libc::c_int, libc::c_int) {
+    let mut fds = [0 as libc::c_int; 2];
+    test_utils::check_system_call!(
+        || { unsafe { libc::pipe2(fds.as_mut_ptr(), libc::O_NONBLOCK) } },
+        &[]
+    )
+    .unwrap();
+    (fds[0], fds[1])
+}
+
+fn pipe_direct_fds_init_helper() -> (libc::c_int, libc::c_int) {
+    let mut fds = [0 as libc::c_int; 2];
+    test_utils::check_system_call!(
+        || { unsafe { libc::pipe2(fds.as_mut_ptr(), libc::O_NONBLOCK | libc::O_DIRECT) } },
+        &[]
+    )
+    .unwrap();
+    (fds[0], fds[1])
+}
+
+fn main() -> anyhow::Result<()> {
+    // should we restrict the tests we run?
+    let filter_shadow_passing = std::env::args().any(|x| x == "--shadow-passing");
+    let filter_libc_passing = std::env::args().any(|x| x == "--libc-passing");
+    // should we summarize the results rather than exit on a failed test
+    let summarize = std::env::args().any(|x| x == "--summarize");
+
+    let all_envs = set![TestEnvironment::Libc, TestEnvironment::Shadow];
+    let mut tests: Vec<test_utils::ShadowTest<(), anyhow::Error>> = vec![];
+
+    let mut add_tests =
+        |name: &str, swappable: bool, fds_init_helper: fn() -> (libc::c_int, libc::c_int)| {
+            // add details to the test names to avoid duplicates
+            let append_args = |s, swapped: bool| {
+                if swappable {
+                    format!("{s} <fds={name:?},swapped={swapped:?}>")
+                } else {
+                    format!("{s} <fds={name:?}>")
+                }
+            };
+
+            for &swapped in if swappable {
+                &[false, true][..]
+            } else {
+                &[false][..]
+            } {
+                let extend_test = |test: fn(libc::c_int, libc::c_int) -> anyhow::Result<()>| {
+                    move || {
+                        let (fd1, fd2) = fds_init_helper();
+                        let (readfd, writefd) = if swapped { (fd2, fd1) } else { (fd1, fd2) };
+                        test(readfd, writefd)
+                    }
+                };
+                tests.extend(vec![
+                    ShadowTest::new(
+                        &append_args("multi-write", swapped),
+                        extend_test(test_multi_write),
+                        all_envs.clone(),
+                    ),
+                    ShadowTest::new(
+                        &append_args("write-then-read", swapped),
+                        extend_test(test_write_then_read),
+                        set![TestEnvironment::Libc],
+                    ),
```

Comment:
This should work now because of the fix in #3277, so can you try using `all_envs.clone()`?